core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -774,7 +778,11 @@ private[log] class Cleaner(val id: Int,
         val batchRetention: BatchRetention =
           if (batch.hasProducerId && isBatchLastRecordOfProducer)
             BatchRetention.RETAIN_EMPTY
-          else if (discardBatchRecords)
+          else if (batch.nextOffset == upperBoundOffsetOfCleaningRound) {
```

Comment:
@vincent81jiang : It seems that we could improve this a bit. Strictly speaking, we only need to preserve this batch if the nextOffset is the same as the high watermark. This will reduce the number of unnecessary empty batches.